File: 041/eu041.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 41
solution by Kevin Retzke (retzkek@gmail.com), April 2012

Problem 41
==========

We shall say that an n-digit number is pandigital if it makes use of all
the digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital
and is also prime.

What is the largest n-digit pandigital prime that exists?

"""
import math

class primes:
    """
    Calculates prime numbers sequentially.

    METHODS

    next()      returns the next prime
    """
    def __init__(self):
        self.primes = [2,3]
        self.primeset = set(self.primes)
        self.lastPrime = 3
    def next(self):
        """
        calculate and return next prime
        """
        nextPrime = None
        i = self.lastPrime+2
        while nextPrime is None:
            sqrt_i = math.sqrt(i)
            isPrime = True
            for p in self.primes:
                if i%p == 0:
                    isPrime = False
                    i += 2
                    break
                if p > sqrt_i:
                    break
            if isPrime:
                nextPrime = i
        self.primes.append(nextPrime)
        self.primeset.add(nextPrime)
        self.lastPrime = nextPrime
        return nextPrime

if __name__ == "__main__":
    # calculate pandigital sets
    pandigital = {1:"1"}
    for i in range(2,10):
        pandigital[i] = pandigital[i-1] + str(i)
    for i in range(1,10):
        pandigital[i] = set(pandigital[i])
    # calculate primes
    p = primes()
    while p.lastPrime < 1e10:
        p.next()
    #for n in xrange(54321,0,-1):
    #    nstr = str(n)
    #    if n in p.primeset and set(nstr) == pandigital[len(nstr)]:
    #        print n
        

File: 041/eu041.go
// Project Euler (projecteuler.net) problem 41
// Kevin Retzke (retzkek@gmail.com), May 2012
package main

import (
    "fmt"
    "math"
)

// Primes is a type for containing a list of primes.
type Primes struct {
	Primes []int
	Last   int
}

// Init initializes a Primes struct with the first two primes.
func (p *Primes) Init() {
	p.Primes = []int{2, 3}
	p.Last = 3
}

// Next computes, returns, and appends the next prime number.
// Computed via trial division by primes.
func (p *Primes) Next() int {
	next := 0
	i := p.Last + 2
	for next == 0 {
		sqrti := math.Sqrt(float64(i))
		isPrime := true
		for _, p := range p.Primes {
			if i%p == 0 {
				isPrime = false
				i += 2
				break
			}
			if float64(p) > sqrti {
				break
			}
		}
		if isPrime {
			next = i
		}
	}
	p.Primes = append(p.Primes, next)
	p.Last = next
	return next
}

// Eratosthenes populates p with all primes up to max, computed with
// the Sieve of Eratosthenes.
func (p *Primes) Eratosthenes(max int) {
    sieve := make([]bool, max)
    for i := 2; i*i < max; i++ {
        if !sieve[i] {
            for j := 2*i; j < max; j += i {
                sieve[j] = true
            }
        }
    }
	p.Primes = []int{2}
    for i := 3; i < max; i++ {
        if !sieve[i] {
            p.Last = i
            p.Primes = append(p.Primes, p.Last)
        }
    }
}

func IsPandigital(n int) bool {
    // determine order of magnitude of n
    mag := 0
    for nn := n; nn > 0; nn = nn/10 {
        mag++
    }
    // find digits in n
    digits := make([]bool,mag)
    for nn := n; nn > 0; nn = nn/10 {
        i := nn%10-1
        if i == -1 {
            return false
        }
        if i < mag && !digits[i] {
            digits[i] = true
        }
    }
    // see if all digits 1:len(n) are in n
    for _,d := range digits {
        if !d {
            return false
        }
    }
    return true
}

func main() {
    primes := new(Primes)
    primes.Eratosthenes(987654321)
    max := 0
    for _,p := range primes.Primes {
        if IsPandigital(p) {
            //fmt.Println(p)
            max = p
        }
    }
    fmt.Println(max)
}
